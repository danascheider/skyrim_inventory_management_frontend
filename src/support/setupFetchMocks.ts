import { vi } from 'vitest'
import createFetchMock from 'vitest-fetch-mock'

/*
 *
 * API Mock Setup
 *
 */

const fetchMocker = createFetchMock(vi)

fetchMocker.enableMocks()
