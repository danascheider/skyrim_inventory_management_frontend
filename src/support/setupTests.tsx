import { type ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { render } from '@testing-library/react'
import { LoginContext } from '../contexts/loginContext'
import { User } from 'firebase/auth'
import testProfileImg from './testProfileImg.png'

export const requireLogin = () => { /* noop */ }

export const testUser = {
  uid: 'edna',
  displayName: 'Edna St. Vincent Millay',
  email: 'edna@gmail.com',
  photoURL: testProfileImg,
  emailVerified: true,
  getIdToken: () => new Promise<string>((resolve, _reject) => resolve('xxxxxxx'))
} as User

export const renderWithRouter = (ui: ReactElement) =>
  render(<BrowserRouter>{ui}</BrowserRouter>)

export const renderAuthenticated = (
  ui: ReactElement,
  authLoading = false,
  authError = undefined
) =>
  renderWithRouter(
    <LoginContext.Provider value={{ user: testUser, token: 'xxxxxxx', requireLogin, authLoading, authError }}>
      {ui}
    </LoginContext.Provider>
  )

export const renderUnauthenticated = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider value={{ user: null, token: null, authLoading: false, requireLogin }}>
      {ui}
    </LoginContext.Provider>
  )

export const renderAuthLoading = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider value={{ user: null, token: null, authLoading: true, requireLogin }}>
      {ui}
    </LoginContext.Provider>
  )
