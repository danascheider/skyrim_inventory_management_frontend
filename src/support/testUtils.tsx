import { type ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { JSDOM } from 'jsdom'
import { render } from '@testing-library/react'
import { LoginContext } from '../contexts/loginContext'
import { User } from 'firebase/auth'
import testProfileImg from './testProfileImg.png'

export const BASE_APP_URI = 'http://localhost:5173'

declare global {
  namespace NodeJS {
    interface Global {
      document: Document
      window: Window
    }
  }
}

const setDom = () => {
  const dom = new JSDOM('<!doctype html><html><body></body></html>', {
    url: 'http://localhost:5173',
  })

  global.window = dom.window as unknown as Window & typeof globalThis
  global.document = dom.window.document
}

/*
 *
 * Global Test Data
 *
 */

export const requireLogin = () => {
  /* noop */
}

export const testUser = {
  uid: 'edna',
  displayName: 'Edna St. Vincent Millay',
  email: 'edna@gmail.com',
  photoURL: testProfileImg,
  emailVerified: true,
  getIdtoken: () =>
    new Promise<string>((resolve, _reject) => resolve('xxxxxxx')),
} as unknown as User

/*
 *
 * Test Renderers
 *
 */

const renderWithJSDOM = (ui: ReactElement) => {
  setDom()

  return render(ui)
}

export const renderWithRouter = (ui: ReactElement) =>
  renderWithJSDOM(<BrowserRouter>{ui}</BrowserRouter>)

export const renderAuthenticated = (ui: ReactElement, authLoading = false) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: testUser,
        token: 'xxxxxxx',
        requireLogin,
        authLoading,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderUnauthenticated = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: null,
        token: null,
        authLoading: false,
        requireLogin,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderAuthLoading = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{ user: null, token: null, authLoading: true, requireLogin }}
    >
      {ui}
    </LoginContext.Provider>
  )
