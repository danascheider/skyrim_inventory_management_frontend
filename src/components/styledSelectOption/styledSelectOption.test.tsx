import { type KeyboardEventHandler, type MouseEventHandler } from 'react'
import { describe, test, expect } from 'vitest'
import { act, fireEvent } from '@testing-library/react'
import { render } from '../../support/testUtils'
import StyledSelectOption from './styledSelectOption'

describe('StyledSelectOption', () => {
  describe('displaying the option', () => {
    describe('when not selected', () => {
      describe('when the option name is shorter than 24 characters', () => {
        test('displays the option name and has aria-selected set to "false"', () => {
          const wrapper = render(
            <StyledSelectOption
              optionName="Option 1"
              optionValue={22}
              onClick={() => {}}
              onKeyDown={() => {}}
              ariaSelected={false}
            />
          )

          expect(wrapper).toBeTruthy()

          const option = wrapper.getByRole('option', { selected: false })

          expect(option).toBeTruthy()
          expect(option.textContent).toEqual('Option 1')
        })

        test('matches snapshot', () => {
          const wrapper = render(
            <StyledSelectOption
              optionName="Option 1"
              optionValue={22}
              onClick={() => {}}
              onKeyDown={() => {}}
              ariaSelected={false}
            />
          )

          expect(wrapper).toMatchSnapshot()
        })
      })

      describe('when the option name is longer than 24 characters', () => {
        test('truncates the name to 24 characters', () => {
          const wrapper = render(
            <StyledSelectOption
              optionName="Neque porro quisquam est quis dolorem ipsum quia dolor sit amet"
              optionValue="De Finibus"
              onClick={() => {}}
              onKeyDown={() => {}}
              ariaSelected={false}
            />
          )

          const option = wrapper.getByRole('option', { selected: false })

          expect(option).toBeTruthy()
          expect(option.textContent).toEqual('Neque porro quisquam es...')
        })

        test('matches snapshot', () => {
          const wrapper = render(
            <StyledSelectOption
              optionName="Neque porro quisquam est quis dolorem ipsum quia dolor sit amet"
              optionValue="De Finibus"
              onClick={() => {}}
              onKeyDown={() => {}}
              ariaSelected={false}
            />
          )

          expect(wrapper).toMatchSnapshot()
        })
      })
    })

    describe('when selected', () => {
      test('displays the option name and has aria-selected set to "false"', () => {
        const wrapper = render(
          <StyledSelectOption
            optionName="Option 1"
            optionValue={1}
            onClick={() => {}}
            onKeyDown={() => {}}
            ariaSelected={true}
          />
        )

        expect(wrapper).toBeTruthy()

        const option = wrapper.getByRole('option', { selected: true })

        expect(option).toBeTruthy()
        expect(option.textContent).toEqual('Option 1')
      })

      test('matches snapshot', () => {
        const wrapper = render(
          <StyledSelectOption
            optionName="Option 1"
            optionValue={1}
            onClick={() => {}}
            onKeyDown={() => {}}
            ariaSelected={true}
          />
        )

        expect(wrapper).toMatchSnapshot()
      })
    })
  })

  describe('event handling', () => {
    test('calls the onClick function when clicked', () => {
      const onClick = vitest.fn() as MouseEventHandler

      const wrapper = render(
        <StyledSelectOption
          optionName="Option 1"
          optionValue="foo"
          onClick={onClick}
          onKeyDown={() => {}}
          ariaSelected={false}
        />
      )

      const option = wrapper.getByRole('option')

      act(() => option.click())

      expect(onClick).toHaveBeenCalledOnce()
    })

    test('calls the onKeyDown function when a key is pressed over it', () => {
      const onKeyDown = vitest.fn() as KeyboardEventHandler

      const wrapper = render(
        <StyledSelectOption
          optionName="Option 1"
          optionValue={1}
          onClick={() => {}}
          onKeyDown={onKeyDown}
          ariaSelected={false}
        />
      )

      const option = wrapper.getByRole('option')

      act(() => fireEvent.keyDown(option))

      expect(onKeyDown).toHaveBeenCalledOnce()
    })
  })
})
