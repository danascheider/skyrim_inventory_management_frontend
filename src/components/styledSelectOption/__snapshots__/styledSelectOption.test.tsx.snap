// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StyledSelectOption > displaying the option > when not selected > when the option name is longer than 24 characters > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        aria-selected="false"
        class="_root_b75fe4"
        data-option-value="De Finibus"
        role="option"
        tabindex="0"
      >
        Neque porro quisquam es...
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      aria-selected="false"
      class="_root_b75fe4"
      data-option-value="De Finibus"
      role="option"
      tabindex="0"
    >
      Neque porro quisquam es...
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`StyledSelectOption > displaying the option > when not selected > when the option name is shorter than 24 characters > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        aria-selected="false"
        class="_root_b75fe4"
        data-option-value="22"
        role="option"
        tabindex="0"
      >
        Option 1
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      aria-selected="false"
      class="_root_b75fe4"
      data-option-value="22"
      role="option"
      tabindex="0"
    >
      Option 1
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`StyledSelectOption > displaying the option > when selected > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        aria-selected="true"
        class="_root_b75fe4"
        data-option-value="1"
        role="option"
        tabindex="0"
      >
        Option 1
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      aria-selected="true"
      class="_root_b75fe4"
      data-option-value="1"
      role="option"
      tabindex="0"
    >
      Option 1
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
